### PR TITLE
Blog app - integration specs for views

### PR DESCRIPTION
**Pull request by r-ahmed2022**
In this PR, we completed the following:-
-  Resolved N+1 problem when fetching all posts and their comments for a user, screenshot attached
Before :- 
image

After using includes
image

-  Used Capybara to write integration tests for each view in our blogsphere app
-  The tests included were home page i.e users listing  ,  individual user page,  listing of posts of a user , and testing individual post

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -4,6 +4,8 @@ class PostsController < ApplicationController
   def index
     @user = User.find(params[:user_id])
     @posts = @user.posts
+    @likes = @posts.includes(:likes)
+    @comments = @posts.includes(:comments)
```
Comment (mckent05) [inline app/controllers/posts_controller.rb]: - Good job done so far, but I think you should make this code a one-liner, making sure that all comments belonging to a post are fetched alongside the post. Please fix this 🙏. Line 6 should be:
`@posts = @user.posts.includes(:comments)`
- Also, according to the project requirements, please add a screenshot of your log before and after the fix to your PR description 👍 👍

File: spec/features/user_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      expect(page).to have_css("img[src='#{@user1.photo}']")
+    end
+
+    it 'shows the number of posts for each user' do
+      expect(page).to have_content(@user1.postscounter)
+    end
+
+    it 'redirects to a user show page when a user is clicked on' do
+      click_link(@user1.name)
+      expect(page).to have_content(@user1.name)
```
Comment (mckent05) [inline spec/features/user_index_spec.rb:24]: **OPTIONAL** - Good job done so far, but to test if the page redirects to the show page, I suggest you test for the current path of the page. Please fix this 😄. E.g:
 `expect(page).to have_current_path user_path(user_id)`, where `user_path` represents the path for your `user#show` page

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    expect(page).to have_content(@comment1.text)
+    expect(page).to have_content(@comment2.text)
+  end
+
+  it 'shows the number of comments and likes for each post' do
+    expect(page).to have_content(@post1.commentscounter)
+    expect(page).to have_content(@post4.commentscounter)
+    expect(page).to have_content(@post2.likescounter)
+    expect(page).to have_content(@post3.likescounter)
+  end
```
Comment (mckent05) [inline spec/features/post_index_spec.rb:47]: **OPTIONAL** - Good job done so far, but to help the readability of your test, please test for the comments_counter and likes_counter in different test blocks 👍 👍

File: spec/features/post_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it 'shows the body of the post' do
+    expect(page).to have_content(@post.text)
+  end
+
+  it 'shows the username of a commenter and their comment' do
+    expect(page).to have_content(@comment.author.name)
+    expect(page).to have_content(@comment.text)
+  end
```
Comment (mckent05) [inline spec/features/post_show_spec.rb:29]: OPTIONAL - Good job done so far, but to help the readability of your test, please test for the name and text in different test blocks 👍 👍

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @r-ahmed2022, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- Good job solving the N+1 query and adding integration tests 👏 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (r-ahmed2022) [reply app/controllers/posts_controller.rb]: Done, as advised

File: spec/features/user_index_spec.rb (hunk 2 above)
Comment (r-ahmed2022) [reply spec/features/user_index_spec.rb:24]: Noted..

File: spec/features/post_index_spec.rb (hunk 3 above)
Comment (r-ahmed2022) [reply spec/features/post_index_spec.rb:47]: Noted, would do in the nexrt PR

File: spec/features/post_show_spec.rb (hunk 4 above)
Comment (r-ahmed2022) [reply spec/features/post_show_spec.rb:29]: Noted, would apply in the next PR

**Review by Reem-lab — APPROVED** [review]
# Hi @r-ahmed2022,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ]  No Linters errors 🔥
- [ ]  Great job working on functionality for tests💯
- [ ]  Followed GitFlow ⭐🚀
- [ ]  Descrtipive PR⭐
- [ ] N+1 problem is fixed ✅



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**